frex/utils/constraint_solver.py

```diff
@@ -0,0 +1,208 @@
+        solution based on the Candidates' total_score (which should be computed by a pipeline) and other constraints.
+
+        :param candidates: A tuple of Candidate objects, with corresponding domain_objects and scores
+        :return: self, with an updated list of candidates
+        """
+        self.candidates = candidates
+        return self
+
+    def set_sections(self, *, num_sections: int):
+        """
+        Set the number of sections in the solution. A section can be thought of as a grouping of items, and the solution
+        will give a set of sections each containing some number of items based on all other constraints and the
+        optimization target.
+        For example, if we are trying to produce a meal plan for 7 days, we can model this by applying the constraint
+        solver with 7 sections (days) where each section contains some number of items (i.e. number of meals per day).
+        Candidates are also expected to have already been appropriately filtered by the pipeline so that we do not
+        consider undesirable candidates during the optimization process.
+
+        :param num_sections: The number of sections for the final solution to contain
+        :return: self, with an updated sections parameter
+        """
+        self.sections = num_sections
+        return self
+
+    def set_items_per_section(self, *, count: int):
+        """
+        Set the number of items that should be assigned to each section in the solution. This corresponds to
+        the number of Candidates that are chosen in the final solution.
+        The implementation currently enforces that each section will contain *exactly* this many items.
+
+        :param count: The number of items that must be assigned to each section.
+        :return: self, with the per_section_count updated
+        """
+        self.per_section_count = count
+        return self
+
+    def add_section_constraint(self, *, attribute_name: str, constraint_type: ConstraintType, constraint_val: int):
+        """
+        Add a constraint to be applied to each section solution. E.g., a constraint on the cost of all items chosen
+        within each given section.
+
+        :param attribute_name: The domain_object's attribute to apply the constraint to
+        :param constraint_type: The type of constraint - i.e. ==, <=, or >=
+        :param constraint_val: The value to constraint the solution to
+        :return: self, with a new Constraint added to the section_constraints list
+        """
+        self.section_constraints.append(
+            Constraint(attribute_name=attribute_name,
+                       constraint_type=constraint_type,
+                       constraint_val=constraint_val)
+        )
+        return self
+
+    def add_overall_constraint(self, *, attribute_name: str, constraint_type: ConstraintType, constraint_val: int) -> ConstraintSolver:
```

Comment:
You usually can't reference the class itself in the method type hints like this, I'm surprised it worked.

Reply:
the import `from __future__ import annotations` allows for postponed evaluation of annotations for python 3.7
https://www.python.org/dev/peps/pep-0563/
apparently 3.8 onwards this works by default without that import statement.